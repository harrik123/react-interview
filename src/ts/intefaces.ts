export interface ITodoItem {
  id: number;
  name: string;
  complete: boolean;
}
