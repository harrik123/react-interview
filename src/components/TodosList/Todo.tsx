import React from "react";
import styled from "styled-components";
import { ITodoProps } from "../../ts/intefaces";
import Button from "../Button";

const Li = styled.li<{ complete: boolean }>`
  display: grid;
  grid-template-columns: 1.5rem 1fr 1.5rem;
  align-items: center;
  padding: 15px;
  margin-bottom: 5px;
  background-color: ${(props) => (props.complete ? "#9EC41C" : "#DD5F5F")};
  color: #fff;
`;

const StatusButton = styled(Button)`
  padding: 0;
  background-color: transparent;
`;

const Name = styled.p`
  margin: 0 15px;
`;

const Incomplete = styled.span`
  display: inline-block;
  height: 1rem;
  width: 1rem;
  border-radius: 50%;
  background-color: #fff;
`;

const RemoveButton = styled(Button)`
  background-color: transparent;
  color: #444;
  padding: 0.25rem;

  &:hover {
    color: #af0000;
  }
`;

function Todo(props: ITodoProps) {
  return (
    <Li complete={props.todo.complete}>
      <StatusButton onClick={() => props.onStatusChangeClick(props.todo.id)}>
        {props.todo.complete ? "âœ…" : <Incomplete></Incomplete>}
      </StatusButton>
      <Name>{props.todo.name}</Name>
      <RemoveButton onClick={() => props.onRemoveClick(props.todo.id)}>
        {"X"}
      </RemoveButton>
    </Li>
  );
}

export default Todo;
