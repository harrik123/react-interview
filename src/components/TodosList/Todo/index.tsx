import React from "react";
import styled from "styled-components";
import { ITodoProps } from "../../../ts/intefaces";
import Button from "../../Button";

const Li = styled.li<{ complete: boolean }>`
  display: grid;
  grid-template-columns: 1.5rem 1fr 1.5rem;
  align-items: center;
  padding: 15px;
  margin-bottom: 5px;
  background-color: ${(props) => (props.complete ? "#9EC41C" : "#DD5F5F")};
  color: #fff;
`;

const StatusButton = styled(Button)<{ complete: boolean }>`
  padding: 0;
  background-color: ${(props) => (props.complete ? "transparent" : "#fff")};
  color: #fff;
  border-radius: 50%;
  width: 1em;
  height: 1em;
`;

const Name = styled.p`
  margin: 0 15px;
`;

const RemoveButton = styled(Button)`
  background-color: transparent;
  color: #444;
  padding: 0.25rem;

  &:hover {
    color: #af0000;
  }
`;

function Todo(props: ITodoProps) {
  return (
    <Li complete={props.todo.complete}>
      <StatusButton
        complete={props.todo.complete}
        onClick={() => props.onStatusChangeClick(props.todo.id)}
      >
        {props.todo.complete ? "âœ“" : ""}
      </StatusButton>
      <Name>{props.todo.name}</Name>
      <RemoveButton onClick={() => props.onRemoveClick(props.todo.id)}>
        {"X"}
      </RemoveButton>
    </Li>
  );
}

export default Todo;
