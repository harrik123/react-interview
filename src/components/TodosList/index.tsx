import React from "react";
import styled from "styled-components";
import { ITodoItem } from "../../ts/intefaces";
import Todo from "./Todo";

const Ul = styled.ul`
  list-style-type: none;
  padding: 0;
  margin: 0;
`;

function TodosList(props: any) {
  return (
    <Ul>
      {props.todos.map((todo: ITodoItem) => {
        return (
          <Todo
            key={todo.id}
            todo={todo}
            onStatusChangeClick={props.onStatusChangeClick}
            onRemoveClick={props.onRemoveClick}
          />
        );
      })}
    </Ul>
  );
}

export default TodosList;
