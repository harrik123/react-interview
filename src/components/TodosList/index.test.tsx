import React from "react";
import renderer from "react-test-renderer";
import TodosList from ".";
import { ITodoItem } from "../../ts/intefaces";

const initTodos: ITodoItem[] = [
  { id: 1, name: "Go to the supermarket", complete: false },
  { id: 2, name: "Call Alice", complete: true },
  { id: 3, name: "Ask Alice to call Bob", complete: false },
  { id: 4, name: "Do the dishes", complete: true },
  { id: 5, name: "Change car tyres", complete: false },
];

describe("<TosdosList />", () => {
  test("TodosList snapshot matches", () => {
    const component = renderer.create(
      <TodosList
        todos={initTodos}
        onRemoveClick={() => {}}
        onStatusChangeClick={() => {}}
      />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
