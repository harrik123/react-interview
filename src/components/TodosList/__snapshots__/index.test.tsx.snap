// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TosdosList /> TodosList snapshot matches 1`] = `
<ul
  className="sc-gKsewC iVAJSz"
>
  <li
    className="sc-gsTCUz kHrTBI"
  >
    <button
      className="sc-bdfBwQ sc-dlfnbm jIvAUB ksHEbo"
      onClick={[Function]}
    >
      <span
        className="sc-eCssSg glwnAj"
      />
    </button>
    <p
      className="sc-hKgILt eNHMqi"
    >
      Go to the supermarket
    </p>
    <button
      className="sc-bdfBwQ sc-jSgupP jIvAUB dtSVeq"
      onClick={[Function]}
    >
      X
    </button>
  </li>
  <li
    className="sc-gsTCUz fEkioF"
  >
    <button
      className="sc-bdfBwQ sc-dlfnbm jIvAUB ksHEbo"
      onClick={[Function]}
    >
      ✅
    </button>
    <p
      className="sc-hKgILt eNHMqi"
    >
      Call Alice
    </p>
    <button
      className="sc-bdfBwQ sc-jSgupP jIvAUB dtSVeq"
      onClick={[Function]}
    >
      X
    </button>
  </li>
  <li
    className="sc-gsTCUz kHrTBI"
  >
    <button
      className="sc-bdfBwQ sc-dlfnbm jIvAUB ksHEbo"
      onClick={[Function]}
    >
      <span
        className="sc-eCssSg glwnAj"
      />
    </button>
    <p
      className="sc-hKgILt eNHMqi"
    >
      Ask Alice to call Bob
    </p>
    <button
      className="sc-bdfBwQ sc-jSgupP jIvAUB dtSVeq"
      onClick={[Function]}
    >
      X
    </button>
  </li>
  <li
    className="sc-gsTCUz fEkioF"
  >
    <button
      className="sc-bdfBwQ sc-dlfnbm jIvAUB ksHEbo"
      onClick={[Function]}
    >
      ✅
    </button>
    <p
      className="sc-hKgILt eNHMqi"
    >
      Do the dishes
    </p>
    <button
      className="sc-bdfBwQ sc-jSgupP jIvAUB dtSVeq"
      onClick={[Function]}
    >
      X
    </button>
  </li>
  <li
    className="sc-gsTCUz kHrTBI"
  >
    <button
      className="sc-bdfBwQ sc-dlfnbm jIvAUB ksHEbo"
      onClick={[Function]}
    >
      <span
        className="sc-eCssSg glwnAj"
      />
    </button>
    <p
      className="sc-hKgILt eNHMqi"
    >
      Change car tyres
    </p>
    <button
      className="sc-bdfBwQ sc-jSgupP jIvAUB dtSVeq"
      onClick={[Function]}
    >
      X
    </button>
  </li>
</ul>
`;
